### PR TITLE
Issue 4387 gl tf

### PR DESCRIPTION
**Pull request by Brijendra21**
Pull request overview
---------------------

Enhancement : OpenStudio could export a web standard GLTF

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/utilities/core/Enum.hpp [hunk 1]
```diff
@@ -75,9 +75,8 @@
   OPENSTUDIO_ENUM_BUILD_STRINGS_PART(_enum_name :: BOOST_PP_SEQ_ELEM(0, _elem), _elem, 1)
```
[the remark this answers is not in the document]
Comment (Brijendra21) [reply src/utilities/core/Enum.hpp:75]: I'm assuming Clang introduced that change..

File: src/model/test/GltfForwardTranslator_GTest.cpp [hunk 2]
```diff
@@ -0,0 +1,512 @@
+/***********************************************************************************************************************
+*  OpenStudio(R), Copyright (c) 2008-2021, Alliance for Sustainable Energy, LLC, and other contributors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (Brijendra21) [reply src/model/test/GltfForwardTranslator_GTest.cpp:2]: Done

File: src/model/test/GltfForwardTranslator_GTest.cpp [hunk 3]
```diff
@@ -43,7 +43,7 @@
 #include "../../osversion/VersionTranslator.hpp"
 #include "../Construction.hpp"
 
-#include "../tiny_gltf.h"
+//#include "../tiny_gltf.h"
```
[the remark this answers is not in the document]
Comment (Brijendra21) [reply src/model/test/GltfForwardTranslator_GTest.cpp:46]: Done

File: src/model/test/GltfForwardTranslator_GTest.cpp [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+#include "../Construction.hpp"
+
+//#include "../tiny_gltf.h"
+
+#include <algorithm>
+
+using namespace openstudio;
+using namespace openstudio::model;
+
+/// <summary>
```
[the remark this answers is not in the document]
Comment (Brijendra21) [reply src/model/test/GltfForwardTranslator_GTest.cpp:53]: Removed /// visual studio type comments and comment tags to plain and simple //.
Looking into the latter part...

File: src/model/GltfForwardTranslator.hpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+   private:
+    //tinygltf::Model gltf_Model;
+    REGISTER_LOGGER("openstudio.model.GltfForwardTranslator");
+
+    StringStreamLogSink m_logSink;
+  };
+
+}  // namespace model
+}  // namespace openstudio
+#endif  //MODEL_GLTFFORWARDTRANSLATOR_HPP
```
[the remark this answers is not in the document]
Comment (Brijendra21) [reply src/model/GltfForwardTranslator.hpp:82]: Added..

File: src/model/GltfForwardTranslator.hpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
+/***********************************************************************************************************************
+*  OpenStudio(R), Copyright (c) 2008-2021, Alliance for Sustainable Energy, LLC, and other contributors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (Brijendra21) [reply src/model/GltfForwardTranslator.hpp:2]: Done..

File: src/model/GltfForwardTranslator.cpp [hunk 7]
```diff
@@ -0,0 +1,1452 @@
+/***********************************************************************************************************************
+*  OpenStudio(R), Copyright (c) 2008-2021, Alliance for Sustainable Energy, LLC, and other contributors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (Brijendra21) [reply src/model/GltfForwardTranslator.cpp:2]: Done

File: src/model/GltfForwardTranslator.cpp [hunk 8]
```diff
@@ -0,0 +1,1452 @@
...
+#include <cmath>
+#include <stack>
+
+namespace openstudio {
+namespace model {
+
+  namespace GLTF = tinygltf;
+  using namespace GLTF;
+
+  /// <summary>
```
[the remark this answers is not in the document]
Comment (Brijendra21) [reply src/model/GltfForwardTranslator.cpp:92]: took care of it..

File: src/model/GltfForwardTranslator.cpp [hunk 9]
```diff
@@ -0,0 +1,1452 @@
...
+  /// Creates and returns a GLTF material
+  /// </summary>
+  /// <param name="name"></param>
+  /// <param name="red"></param>
+  /// <param name="green"></param>
+  /// <param name="blue"></param>
+  /// <param name="alpha"></param>
+  /// <param name="isDoubleSided"></param>
+  /// <returns>GLTF Material</returns>
+  GLTF::Material CreateMaterial(const std::string& name, float red, float green, float blue, float alpha, bool isDoubleSided) {
```
[the remark this answers is not in the document]
Comment (Brijendra21) [reply src/model/GltfForwardTranslator.cpp:126]: done.

File: src/model/GltfForwardTranslator.cpp [hunk 10]
```diff
@@ -0,0 +1,1452 @@
...
+  std::vector<mat_Data> GetAllMaterials() {
+    std::vector<mat_Data> result;
+
+    result.push_back({"Undefined", 255, 255, 255, 1, true});
+
+    result.push_back({"NormalMaterial", 255, 255, 255, 1, true});
+    result.push_back({"NormalMaterial_Ext", 255, 255, 255, 1});
+    result.push_back({"NormalMaterial_Int", 255, 0, 0, 1});
+
+    //result.push_back({"Floor", 128, 128, 128, 1, true});
```
[the remark this answers is not in the document]
Comment (Brijendra21) [reply src/model/GltfForwardTranslator.cpp:257]: done

File: src/model/GltfForwardTranslator.hpp [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+    /// Convert an OpenStudio Model to Gltf format
+    bool modelToGLTF(const Model& model, const path& outputPath);
+    bool modelToGLTF(const Model& model, std::function<void(double)> updatePercentage, const path& outputpath);
+
+    //tinygltf::Model to json
+    bool toJSON();
+
+    //Create Triangle gltf
+    bool CreateTriangleGLTF(const path& outputPath);
```
[the remark this answers is not in the document]
Comment (Brijendra21) [reply src/model/GltfForwardTranslator.hpp]: made changes accordingly.

File: src/model/test/GltfForwardTranslator_GTest.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Brijendra21) [reply src/model/test/GltfForwardTranslator_GTest.cpp]: C++ gltf validator's status is still experimental (W.I.P). Almost every rendering platform provides the validation results as soon as we try rendering a gltf so I think this isn't an issue.

File: src/model/GltfForwardTranslator.cpp [hunk 12]
```diff
@@ -0,0 +1,1330 @@
...
+  }
+
+  // Adds Normal Buffers for all normal Vectors
+  // param : normalVectors
+  // param : _coordinatesBuffer
+  // param : _accessors
+  // returns : index for the Normals Buffer
+  int addNormals(Vector3dVector& normalVectors, std::vector<unsigned char>& _coordinatesBuffer, std::vector<GLTF::Accessor>& _accessors) {
+    std::vector<float> values;
+    float divider = 1;
```
Comment (jmarrec) [inline src/model/GltfForwardTranslator.cpp:718]: What's the point of this divider = 1?

**Review by jmarrec — CHANGES_REQUESTED** [review]
There are portion of the code that I'm not qualified/smart enough to fully understand but there are a few C++ improvements and code smell I'm seeing.

As this is a rather big amount of review comments, I'll going to a make a PR with some refactoring of this, and I'll let you take a look at it before you merge (partially or fully) into this branch.

File: src/model/GltfForwardTranslator.cpp [hunk 13]
```diff
@@ -0,0 +1,1330 @@
...
+    }
+    return bytes;
+  }
+
+  // Adds & Creates Face Indices buffers and Accessors
+  // param : faceIndices
+  // param : _indicesBuffer
+  // param : _accessors
+  // returns : index of the Face Indices
+  int addIndices(std::vector<size_t>& faceIndices, std::vector<unsigned char>& _indicesBuffer, std::vector<GLTF::Accessor>& _accessors) {
```
[the remark this answers is not in the document]
Comment (Brijendra21) [reply src/model/GltfForwardTranslator.cpp:568]: // Adds & Creates Face Indices buffers and Accessors
  // This method infers the indicesbuffer and adds them to the glTF model accessor
  // which expects two types of accessor input one from indicesbuffer and the second
  // one form coordinates buffer. So here after appending the buffer content it is sending
  // over the index so the containing node will be aware of which one to refer.
  // A better overview here at [Infographic  ](https://github.com/KhronosGroup/glTF/blob/main/specification/2.0/figures/gltfOverview-2.0.0b.png)
  // param : faceIndices
  // param : _indicesBuffer
  // param : _accessors
  // returns : index of the Face Indices

File: src/model/GltfForwardTranslator.cpp [hunk 14]
```diff
@@ -0,0 +1,1330 @@
...
+    _nodesNew.push_back(_topNode);
+    int sk = 0;
+    for (auto& nn : _nodes) {
+      if (sk != 0) {
+        _nodesNew.push_back(nn);
+      }
+      sk++;
+    }
+
+    gltf_Model.nodes = _nodesNew;
```
[the remark this answers is not in the document]
Comment (Brijendra21) [reply src/model/GltfForwardTranslator.cpp:1033]: if I'm changing then top node is not set properly and rendered orientation isn't correct.
image

Originally 
image

[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/GltfForwardTranslator.cpp]: @Brijendra21 Couple of things then:
* This isn't captured in unit tests and probably should (all tests were passing locally when I made my changes)
* we should identify why and correct it, it's probably just a small hiccup. It's a waste to have to create a new vector and push back all elements into it. Just ensure topNode is the first one to begin with.

**Pull request merged**